File: cpp/two_sum.cc
/*
 * Filename: two_sum.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Thu 23 Jul 2015 11:01:12 AM EDT
 */

#include <iostream>
using namespace std;

class Solution
{
	public:
		vector<int> twoSum(vector<int>& nums, int target) {
			vector<int> res;
			map<int, int> map;

			for (int i = 0; i < nums.size(); i++) {
				int n = nums[i];
				int rest = target - n;

				if (map.find(rest) != map.end()) {
					res.push_back(map[rest] + 1);
					res.push_back(i+1);
					return res;
				}

				map[nums[i]] = i;
			}

			return res;
		}
};

File: java/TwoSum.java
/*
 * Filename: TwoSum.java
 * Author: Jun Xu
 * Email: junx@bu.edu
 * Created Time: Tue 03 Mar 2015 12:21:05 AM EST
 */

public class Solution {
	public int[] twoSum(int[] numbers, int target) {
		HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();
		int[] res = new int[2];

		for (int i = 0; i < numbers.length; i++) {
			if (map.containsKey(target - numbers[i])) {
				res[0] = 1 + map.get(target - numbers[i]);
				res[1] = 1 + i;
				return res;
			}
			map.put(numbers[i], i);
		}

		return res;
	}
}
